### DISCUSSION TITLE
Select feature from custom style layer.

### DISCUSSION THREAD

**Discussion by devolod** [maplibre/maplibre-gl-js#2105] [Q&A]
Hi All!

The question about ability to select feature that drawn directly via shaders as in the [example](https://maplibre.org/maplibre-gl-js-docs/example/custom-style-layer/) and get some property value from that feature.
I.e:
```
map.on('click', 'custom_style_layer', function(e) {  
    console.log(e.features[0]['some_property'])
})
```
Is there way to do it?

**Comment by HarelM** [comment]
Can you please attach a jsbin/codepen/stackblitz initial code with a custom layer and maybe some instructions how to use this custom layer and what you expect to be able to do?
Generally speaking you can use querysourcefeature to get data from a source, queryrender is for "regularly" rendered features I believe.

**Reply by devolod** [reply]
Final goal is to draw 3d lines ( gas pipeline and other communications ) with following abilities:

 - get properties of a line by click
 - change style of a selected line
 - show tooltip on mouse hover would be great also
 
 I started with simple 2d line since drawing with altitude is complex WebGL task.
Here [codepen](https://codepen.io/userV_F/pen/GRBBYWy).

**Reply by HarelM** [reply]
3D lines are currently not supported in maplibre.
You can probably do a custom layer, but I'm guessing this won't be easy...

**Comment by devolod** [comment]
> Generally speaking you can use querysourcefeature to get data from a source

I tried `querySourceFeatures`, but didn't succeed. 
By the way, `map.getStyle().layers` returns empty array, maybe it's related.
[Codepen](https://codepen.io/userV_F/pen/GRBBYWy)  with details.

**Reply by HarelM** [reply]
Query source works on rendered layers I think (even hidden ones), not super intuitive...
